File: src/main/java/com/catboyranch/roborancher/managers/RoleMessageManager.java
package com.catboyranch.roborancher.managers;

import com.catboyranch.roborancher.Server;
import com.catboyranch.roborancher.utils.ChannelUtils;
import com.catboyranch.roborancher.utils.RoleUtils;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.emoji.Emoji;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.HashMap;

public class RoleMessageManager {
    private final Server server;
    private final HashMap<String, RoleMessage> roleMessages = new HashMap<>();

    public RoleMessageManager(Server server) {
        this.server = server;
    }

    public void load(JSONObject json) {
        if(json.has("roleMessages")) {
            JSONArray roleMessagesJSON = json.getJSONArray("roleMessages");
            for (int i = 0; i < roleMessagesJSON.length(); i++) {
                JSONObject rmJSON = roleMessagesJSON.getJSONObject(i);
                roleMessages.put(rmJSON.getString("id"), new RoleMessage(rmJSON));
            }
        }
    }

    public void addRoleMessageEmoji(Message message, Emoji emoji, Role role) {
        String messageID = message.getId();
        if(!roleMessages.containsKey(messageID))
            roleMessages.put(messageID, new RoleMessage(messageID));
        roleMessages.get(messageID).addReaction(emoji.getFormatted(), role.getId());
    }

    public void ensureRoleMessageEmojis() {
        for(String messageID : roleMessages.keySet()) {
            ChannelUtils.getMessage(server, messageID, objects -> {
                if(objects.length == 0) {
                    System.out.println("Error ensuring role message emoji for message id " + messageID);
                    roleMessages.remove(messageID);
                    return;
                }
                RoleMessage rm = roleMessages.get(messageID);
                Message message = (Message)objects[0];
                for(String emoji : rm.getReactions().keySet())
                    message.addReaction(Emoji.fromFormatted(emoji)).queue((success) -> {}, (failure) -> {
                        //Emoji didnt work, remove it
                        System.out.printf("Could not add emoji %s to message %s!\n", emoji, messageID);
                        rm.removeReaction(emoji);
                    });
            });
        }
    }

    public void removeRoleMessageEmoji(Message message, Emoji emoji) {
        String messageID = message.getId();
        if(!roleMessages.containsKey(messageID))
            return;
        RoleMessage rm = roleMessages.get(messageID);
        rm.removeReaction(emoji.getFormatted());
        if(rm.shouldRemove()) {
            message.removeReaction(emoji).queue();
            roleMessages.remove(messageID);
        }
    }

    public Role getRoleForMessageEmoji(String messageID, Emoji emoji) {
        if(!roleMessages.containsKey(messageID))
            return null;

        HashMap<String, String> map = roleMessages.get(messageID).getReactions();
        for(String currentEmoji : map.keySet()) {
            if(currentEmoji.equals(emoji.getFormatted())) {
                return RoleUtils.getRole(map.get(currentEmoji), server);
            }
        }
        return null;
    }

    public JSONArray toJSON() {
        JSONArray roleMessagesJSON = new JSONArray();
        for(String messageID : roleMessages.keySet()) {
            RoleMessage rm = roleMessages.get(messageID);
            if(!rm.getReactions().isEmpty())
                roleMessagesJSON.put(rm.toJSON());
        }
        return roleMessagesJSON;
    }
}

File: src/main/kotlin/com/catboyranch/roborancher/managers/RoleMessageManager.kt
package com.catboyranch.roborancher.managers

import com.catboyranch.roborancher.ChannelUtils
import com.catboyranch.roborancher.IFunction
import com.catboyranch.roborancher.RoleUtils
import com.catboyranch.roborancher.Server
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.Role
import net.dv8tion.jda.api.entities.emoji.Emoji

class RoleMessageManager(private val server: Server) {
    val rm = server.config.getData().roleMessages

    fun addRoleMessageEmoji(message: Message, emoji: Emoji, role: Role) {
        val messageID = message.id
        if(!rm.containsKey(messageID))
            rm[messageID] = HashMap()
        rm[messageID]?.set(emoji.formatted, role.id)
        message.addReaction(emoji).queue()
    }

    fun removeRoleMessageEmoji(message: Message, emoji: Emoji) {
        val messageID = message.id
        val map = rm[messageID] ?: return
        message.removeReaction(emoji).queue()
        map.remove(emoji.formatted)
        if(map.isEmpty())
            rm.remove(messageID)
    }

    fun getRoleForMessageEmoji(messageID: String, emoji: Emoji): Role? = rm[messageID]?.get(emoji.formatted)?.let { RoleUtils.getRole(server, it) }

    fun ensureRoleMessageEmojis() {
        rm.forEach { (messageID, roleMessage) ->
            if(roleMessage.isEmpty()) {
                //Role message is empty in data, just delete it and return
                rm.remove(messageID)
                return
            }
            ChannelUtils.getMessage(server, messageID, object: IFunction {
                override fun run(vararg args: Any) {
                    if(args.isEmpty()) {
                        println("Error ensuring role message emoji for message id $messageID")
                        rm.remove(messageID)
                        return
                    }
                    val message = args[0] as Message
                    rm[messageID]?.forEach { (emoji, _) ->
                        message.addReaction(Emoji.fromFormatted(emoji)).queue({ }, {
                            println("Could not add emoji $emoji to message $messageID")
                            rm[messageID]?.remove(emoji)
                        })
                    }
                }

            })
        }
    }
}